documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+
+    public static Vector2D<TFloat> LerpClamped<T, TFloat>(Vector2D<T> value1, Vector2D<T> value2, TFloat amount) where T : INumberBase<T> where TFloat : INumberBase<TFloat>, IFloatingPoint<TFloat>;
+
+    // internal static Vector2D<T> LerpClampedUnchecked<T>(Vector2D<T> value1, Vector2D<T> value2, T amount) where T : INumberBase<T>;
+
+    /// <summary>Returns a vector whose elements are the maximum of each of the pairs of elements in two specified vectors.</summary>
+    /// <param name="value1">The first vector.</param>
+    /// <param name="value2">The second vector.</param>
+    /// <returns>The maximized vector.</returns>
+    public static Vector2D<T> Max<T>(Vector2D<T> value1, Vector2D<T> value2) where T : INumberBase<T>;
+
+    /// <summary>Returns a vector whose elements are the minimum of each of the pairs of elements in two specified vectors.</summary>
+    /// <param name="value1">The first vector.</param>
+    /// <param name="value2">The second vector.</param>
+    /// <returns>The minimized vector.</returns>
+    public static Vector2D<T> Min<T>(Vector2D<T> value1, Vector2D<T> value2) where T : INumberBase<T>;
+
+    /// <summary>Returns a vector with the same direction as the specified vector, but with a length of one.</summary>
+    /// <param name="value">The vector to normalize.</param>
+    /// <returns>The normalized vector.</returns>
+    public static Vector2D<TReturn> Normalize<T, TReturn>(Vector2D<T> value) where T : INumberBase<T> where TReturn : INumberBase<TReturn>, IRootFunctions<TReturn>;
+
+    /// <summary>Returns a vector with the same direction as the specified vector, but with a length of one.</summary>
+    /// <param name="value">The vector to normalize.</param>
+    /// <returns>The normalized vector.</returns>
+    public static Vector2D<T> Normalize<T>(Vector2D<T> value) where T : INumberBase<T>, IRootFunctions<T>;
+
+    /// <summary>Returns the reflection of a vector off a surface that has the specified normal.</summary>
+    /// <param name="vector">The source vector.</param>
+    /// <param name="normal">The normal of the surface being reflected off.</param>
+    /// <returns>The reflected vector.</returns>
+    public static Vector2D<TReturn> Reflect<T, TReturn>(Vector2D<T> vector, Vector2D<T> normal) where T : INumberBase<T> where TReturn : INumberBase<TReturn>;
+
+    /// <summary>Returns a vector whose elements are the square root of each of a specified vector's elements.</summary>
+    /// <param name="value">A vector.</param>
+    /// <returns>The square root vector.</returns>
+    public static Vector2D<TReturn> Sqrt<T, TReturn>(Vector2D<T> value) where T : INumberBase<T> where TReturn : INumberBase<TReturn>, IRootFunctions<TReturn>;
+
+    public static Vector2D<T> Remainder<T>(this Vector2D<T> left, Vector2D<T> right) where T : INumberBase<T>, IModulusOperators<T, T, T>;
```

Comment:
Which remainder is this and does it make sense in the context of all types and when paired with `Divide` in the generic math sense?

Reply:
This creates a vector `a % b` for all components in a and b. It does behave correspondingly to `Divide` and (I think) has the right semantics for IModulusOperators.

Reply:
One thing I’ve learnt through API reviews is that there are apparently different remainders haha. Maths is nuts.

Reply:
Yep. The `%` operator in C# (and C/C++) is functionally remainder of truncated division (most commonly described as `rem`). It behaves as a `modulus` when the signs of both inputs match.

Some people instead expect a behavior commonly referred to as `mod`, which is remainder of floored division. It behaves as a `modulus` regardless of the signs.

For integers, `/` and `%` are a pair with `%` functionally being `x - ((x/ y) * y)`. For floating-point, `/` doesn't return the integer result, but rather the floating-point result and thus you need to do the additional rounding step (`truncate`) to get the pair of `%`.

This can make it difficult to use correctly in some generic contexts and is potentially worth considering how that all works.

-- `IEEE754Remainder` is then the remainder of division that uses `round to nearest, ties to even` to produce its integer based quotient.